homeassistant/components/habitica/__init__.py

```diff
@@ -151,3 +62,55 @@ async def handle_api_call(call):
         DOMAIN, SERVICE_API_CALL, handle_api_call, schema=SERVICE_API_CALL_SCHEMA
     )
     return True
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Set up habitica from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+    config = config_entry.data
+    data = hass.data[DOMAIN]
+    websession = async_get_clientsession(hass)
+
+    class HAHabitipyAsync(HabitipyAsync):
+        """Closure API class to hold session."""
+
+        def __call__(self, **kwargs):
+            return super().__call__(websession, **kwargs)
+
+    url = config[CONF_URL]
+    username = config[CONF_API_USER]
+    password = config[CONF_API_KEY]
+    name = config.get(CONF_NAME)
+    config_dict = {"url": url, "login": username, "password": password}
+    api = HAHabitipyAsync(config_dict)
+    user = await api.user.get()
+    if name is None:
+        name = user["profile"]["name"]
+        hass.config_entries.async_update_entry(
+            config_entry, data={**config_entry.data, CONF_NAME: name},
+        )
+    data[name] = api
```

Comment:
Our standard is to use the `config_entry.entry_id` as key in hass.data for entry specific data.
```suggestion
    data[config_entry.entry_id] = api
```